### ISSUE TITLE
Is there a way to get notified by email when any new post get posted in any forum?

### ISSUE THREAD

**Issue by hopewise** [rubysherpas/forem#566]
Is there a way to get notified by email when any new post get posted in any forum?

**Comment by ncri** [comment]
Ah, so this doesn't have email notifications? Surprising, as it's a basic forum feature.

We are looking into using forem, but without notifications we probably won't. Unless there is an easy way of adding them?

**Comment by radar** [comment]
It has [one subscription notification](https://github.com/radar/forem/blob/rails4/app/mailers/forem/subscription_mailer.rb). The main reason it doesn't have this feature is because:
1. I have a full-time job and in what little spare time I have I'm writing [a book](https://leanpub.com/multi-tenancy-rails) and spending my time on many, many other things. Thus a complete lack of any time to devote to Forem.
2. I'm running this project as a little experiment and trying to encourage people to contribute back to it vs. doing all the work myself. 

If you would like email notifications to be a "thing" within Forem, please submit a patch to add them in.

**Comment by ncri** [comment]
Reply notifications are the most important notifications, I assumed forem doesn't have any notifications. So I think we'll give it a try. ;-) And thx for the book link, looks interesting!

**Comment by hopewise** [comment]
Reply notifications is working, but new post is not sending email to admin, how would I add this feature? what controller / action shall I over ride? 
Also, how would I make posts auto approved rather than getting moderated?
Thanks for Forem!

**Comment by ncri** [comment]
I just tested and it didn't work, but it might be my apps email setup, i will check

**Comment by hopewise** [comment]
make sure that you are subscribed

**Comment by ncri** [comment]
Hm, i am subscribed now and still don't get notified for replies. Also as the post author i would expect to be auto subscribed.

**Comment by ncri** [comment]
Ah, i see now why! Notifications are sent when a post is approved. But I don't use post approval, i set the state of new posts to "approved" by default. @radar is there another way of not having to approve posts then setting database defaults?

**Comment by ncri** [comment]
I tried adding an after create filter to the post model to notify users, but it doesn't get called:

```
Forem::Post.class_eval do
  after_create :email_topic_subscribers
end
```

Any idea why? Adding a method and calling it from a post instance works.

**Comment by ncri** [comment]
Trying to debug this. Quite strange, when I do:

```
Forem::Post.class_eval do
  after_save do
    raise 'test'
  end
end
```

The error is correctly raised in the console when saving a post, but not when creating/saving one via the web interface.

**Comment by ncri** [comment]
Okay, the issue is, that the callback is only added for the first request for some reason, when using rails server. I got it working by doing:

```
Rails.application.config.to_prepare do
  Forem::Post.class_eval do
    after_create :email_topic_subscribers
  end
end
```

**Comment by hopewise** [comment]
so, if I just added the code above as an initialization code in Rails, it
will work?
did you figure how to set a new post automatically approved? currently it
will ask for moderation first ..

On Wed, Aug 13, 2014 at 4:21 PM, Nico Ritsche notifications@github.com
wrote:

> Okay, the issue is, that the callback is only added for the first request
> for some reason, when using rails server. I got it working by doing:
> 
> Rails.application.config.to_prepare do
>   Forem::Post.class_eval do
>     after_create :email_topic_subscribers
>   end
> end
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/radar/forem/issues/566#issuecomment-52046598.

## 

## 

## Kind Regards,

Samir Sabri
Software Architect& Developer
www.dcaclab.com
Jordan-Middle East

**Comment by ncri** [comment]
I did it via a migration that sets the state to approved by default, for various models. It was mentioned in another issue:

```
def up
  change_column_default(:forem_topics, :state, 'approved')
  change_column_default(:forem_posts, :state, 'approved')
  change_column_default(:users, :forem_state, 'approved')
end

def down
  change_column_default(:forem_topics, :state, 'pending_review')
  change_column_default(:forem_posts, :state, 'pending_review')
  change_column_default(:users, :forem_state, 'pending_review')
end
```

But you can also try to do it like this for posts:

```
Rails.application.config.to_prepare do
  Forem::Post.class_eval do
    after_create :approve
  end
end
```

Untested. More elegant and you can avoid having to call email_topic_subscribers separately. You will still have to handle topic and user approval then.

**Comment by hopewise** [comment]
Thanks Nico for your contribution

On Wed, Aug 13, 2014 at 4:32 PM, Nico Ritsche notifications@github.com
wrote:

> I did it via a migration that sets the state to approved by default, for
> various models. It was mentioned in another issue:
> 
> def up
>   change_column_default(:forem_topics, :state, 'approved')
>   change_column_default(:forem_posts, :state, 'approved')
>   change_column_default(:users, :forem_state, 'approved')
> end
> 
> def down
>   change_column_default(:forem_topics, :state, 'pending_review')
>   change_column_default(:forem_posts, :state, 'pending_review')
>   change_column_default(:users, :forem_state, 'pending_review')
> end
> 
> But you can also try to do it like this for posts:
> 
> Rails.application.config.to_prepare do
>   Forem::Post.class_eval do
>     after_create :approve
>   end
> end
> 
> Untested. More elegant and you can avoid having to call
> email_topic_subscribers separately. You will still have to handle topic and
> user approval then.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/radar/forem/issues/566#issuecomment-52047842.

## 

## 

## Kind Regards,

Samir Sabri
Software Architect& Developer
www.dcaclab.com
Jordan-Middle East

**Comment by radar** [comment]
@ncri: If you're going to set all posts to approve like that, why not run it _once_ in a console command and have `Forem.moderate_first_post` set to `false` so that new topics/posts are not marked as needing review?

**Comment by ncri** [comment]
Ah, setting Forem.moderate_first_post to false completely switches off the need for approving posts? Will email notifications still be triggered? And what do you mean I should do in a rake task only once?

**Comment by radar** [comment]
Yes it will turn off the need for approving posts. It doesn't alter the email notifications. 

What would be the point of setting all posts to open when `Forem.moderate_first_post` is `false` more than once? There'd be no posts after that fact to be set to open.

**Comment by ncri** [comment]
Yeah, I'm just confused about what you mean with "it" in "why not run **it** once in a console command". As far as I understand you I just have to set Forem.moderate_first_post to false? Currently in my app the post's state is set to approved by default at database level. That will then not be necessary anymore and actually will prevent notifications. So I will have to change the default back to pending_review?

**Comment by radar** [comment]
This is the "it":

```
Forem.Post.update_all('state', 'approved')
Forem.Topic.update_all('state', 'approved')
```

All posts and all topics will move from 'pending_review' to 'approved'.

**Comment by ncri** [comment]
okay, thx, got it.
